src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.TryCloneFile.OSX.cs

```diff
@@ -0,0 +1,108 @@
+                int destError = Interop.Sys.Stat(destFullPath, out Interop.Sys.FileStatus destStat);
+                if (destError != 0)
+                {
+                    // stat failed. If the destination doesn't exist anymore,
+                    // try clonefile; otherwise, fall back to a normal copy.
+                    if (Interop.Sys.GetLastError() == Interop.Error.ENOENT)
+                    {
+                        goto tryCloneFile;
+                    }
+                    else
+                    {
+                        return false;
+                    }
+                }
+                else
+                {
+                    // Destination exists:
+                    if (srcStat.Dev != destStat.Dev)
+                    {
+                        // On different device, so fall back to normal copy
+                        return false;
+                    }
+                    if (srcStat.Ino == destStat.Ino)
+                    {
+                        // Copying onto itself
+                        throw new IOException(SR.Format(SR.IO_SharingViolation_File, destFullPath));
+                    }
+                }
+
+                // Try deleting destination:
+                {
+                    // Delete the destination. This should fail on directories.
+                    // Get a lock to the dest file to ensure we don't copy onto it when it's locked by something else, and then delete it.
+                    try
+                    {
+                        using SafeFileHandle? dstHandle = SafeFileHandle.Open(destFullPath, FileMode.Open, FileAccess.ReadWrite,
+                            FileShare.None, FileOptions.None, preallocationSize: 0, filePermissions);
+                        File.Delete(destFullPath);
+                    }
+                    catch (FileNotFoundException)
+                    {
+                        // We don't want to throw if it's just the file not existing, since we're trying to delete it.
+                    }
+                }
+
+                // Try clonefile:
+                tryCloneFile:
+                {
+                    if (Interop.@libc.clonefile(sourceFullPath, destFullPath, Interop.@libc.CLONE_ACL) == 0)
+                    {
+                        // Success
+                        return true;
+                    }
+                    error = Interop.Sys.GetLastError();
+                }
+            }
+
+            // Check if it's not supported, or if they're on different filesystems.
+            if (error == Interop.Error.ENOTSUP || error == Interop.Error.EXDEV)
+            {
+                // Fall back to normal copy
+                return false;
+            }
+
+            // Check we didn't get an error due to 'invalid flags' (which should never happen)
+            Debug.Assert(error != Interop.Error.EINVAL);
+            if (error == Interop.Error.EINVAL)
```

Comment:
Can you remove this? There is no reason to have to special-case `EINVAL`.

Reply:
There are test failures related to this :)
I suspect that it's on filesystems without ACLs, so I will add some code to deal with this soon

Reply:
I would assume this is probably APFS with ACLs disabled, but I honestly don't know. Will just add some code that should work though.

Reply:
This is because `clonefile` can in principle fail with any error at any time since it's implemented by each filesystem driver thing. This means that EINVAL could still probably technically happen even if flags were 0, I will account for this possibility.